lib/cretonne/src/loop_analysis.rs

```diff
@@ -0,0 +1,324 @@
+//! A loop analysis respresented as mappings of loops to their header Ebb
+//! and parent in the loop tree.
+
+
+use ir::{Function, Ebb, Layout, Loop};
+use flowgraph::ControlFlowGraph;
+use dominator_tree::DominatorTree;
+use entity_map::{EntityMap, PrimaryEntityData};
+use packed_option::PackedOption;
+use std::collections::HashSet;
+
+/// Loop tree information for a single function.
+///
+/// Loops are referenced by the Loop object, and for each loop you can access its header EBB,
+/// its eventual parent in the loop tree and all the EBB belonging to the loop.
+pub struct LoopAnalysis {
+    loops: EntityMap<Loop, LoopData>,
+    ebbs: EntityMap<Ebb, PackedOption<Loop>>,
+}
+
+struct LoopData {
+    header: Ebb,
+    parent: PackedOption<Loop>,
+}
+
+impl PrimaryEntityData for LoopData {}
+
+/// Methods for querying the loop analysis.
+impl LoopAnalysis {
+    /// Allocate a new blank loop analysis struct. Use `compute` to compute the loop analysis for
+    /// a function.
+    pub fn new() -> LoopAnalysis {
+        LoopAnalysis {
+            loops: EntityMap::new(),
+            ebbs: EntityMap::new(),
+        }
+    }
+
+    /// Returns all the loops contained in a function.
+    pub fn get_loops(&self) -> Vec<Loop> {
```

Comment:
In fact I need to iterate the `Loop`s in reverse order for the second pass of the loop detections so I would need to implement `std::iter::DoubleEndedIterator` for `entity_map::Keys`.